bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/TypeChecker.java

```diff
@@ -1256,83 +1310,123 @@ private static boolean checkIsArrayType(Type sourceType, BArrayType targetType,
         }
+        return checkIsArrayType((BTupleType) sourceType, targetType, unresolvedTypes);
+    }
 
-        switch (sourceArrayType.getState()) {
-            case UNSEALED:
-                if (targetType.getState() != ArrayType.ArrayState.UNSEALED) {
+    private static boolean checkIsTupleType(BArrayType sourceType, BTupleType targetType,
+                                            List<TypePair> unresolvedTypes) {
+        Type sourceElementType = sourceType.getElementType();
+        List<Type> targetTypes = new ArrayList<>(targetType.getTupleTypes());
+        Type targetRestType = targetType.getRestType();
+
+        switch (sourceType.getState()) {
+            case OPEN:
+                if (targetRestType == null) {
                     return false;
                 }
-                break;
-            case CLOSED_SEALED:
-                if (targetType.getState() == ArrayType.ArrayState.CLOSED_SEALED &&
-                        sourceArrayType.getSize() != targetType.getSize()) {
+                if (targetTypes.isEmpty()) {
+                    return checkIsType(sourceElementType, targetRestType, unresolvedTypes);
+                }
+                return false;
+            case CLOSED:
+                if (sourceType.getSize() < targetTypes.size()) {
                     return false;
                 }
-                break;
-        }
+                if (targetTypes.isEmpty()) {
+                    if (targetRestType != null) {
+                        return checkIsType(sourceElementType, targetRestType, unresolvedTypes);
+                    }
+                    return sourceType.getSize() == 0;
+                }
 
-        //If element type is a value type, then check same type
-        Type targetElementType = targetType.getElementType();
-        int targetElementTypeTag = targetElementType.getTag();
+                for (Type targetElementType : targetTypes) {
+                    if (!(checkIsType(sourceElementType, targetElementType, unresolvedTypes))) {
+                        return false;
+                    }
+                }
+                if (sourceType.getSize() == targetTypes.size()) {
+                    return true;
+                }
+                if (targetRestType != null) {
+                    return checkIsType(sourceElementType, targetRestType, unresolvedTypes);
+                }
+                return false;
+            default:
+                return false;
+        }
+    }
 
-        Type sourceElementType = sourceArrayType.getElementType();
+    private static boolean checkIsTupleType(BTupleType sourceType, BTupleType targetType,
+                                            List<TypePair> unresolvedTypes) {
+        List<Type> sourceTypes = new ArrayList<>(sourceType.getTupleTypes());
+        Type sourceRestType = sourceType.getRestType();
+        List<Type> targetTypes = new ArrayList<>(targetType.getTupleTypes());
```

Comment:
Here again, for both `sourceTypes` and `targetTypes`, since we are no longer adding the rest types to these lists we can directly use the lists returned by `getTupleTypes()` instead of creating new lists.